File: php/Assembler/src/Loader/LoaderNormal.php
<?php

namespace Assembler\Loader;

use Arshu\Common\Logger;
use Assembler\Common\CommonUtil;

/**
 * Template result structure
 */
class TemplateResult
{
    public string $html;
    public ?string $json;

    public function __construct(string $html, ?string $json = null)
    {
        $this->html = $html;
        $this->json = $json;
    }
}

/**
 * Handles loading and caching of HTML templates from the file system
 */
class LoaderNormal
{
    private static array $htmlTemplatesCache = [];

    /**
     * Loads HTML files and corresponding JSON files from the specified application site directory, caching the output per appSite
     * @param string $rootDirPath Root directory path
     * @param string $appSite Application site name
     * @return array<string, TemplateResult> Array of templates
     */
    public static function loadGetTemplateFiles(string $rootDirPath, string $appSite, string $searchAppSites = ""): array
    {
        Logger::debug("LoadGetTemplateFiles called for appSite: $appSite, searchAppSites: $searchAppSites", 'LoaderNormal');

        $cacheKey = dirname($rootDirPath) . '|' . $appSite . '|' . $searchAppSites;

        if (isset(self::$htmlTemplatesCache[$cacheKey])) {
            $cached = self::$htmlTemplatesCache[$cacheKey];
            Logger::debug("Returning cached templates for $appSite (" . count($cached) . " templates)", 'LoaderNormal');
            return $cached;
        }

        // Load templates from primary appSite
        $result = self::loadTemplatesFromSingleAppSite($rootDirPath, $appSite);

        // Load templates from searchAppSites for fallback
        if (!empty($searchAppSites)) {
            $searchAppSitesArray = explode(',', $searchAppSites);
            foreach ($searchAppSitesArray as $searchAppSiteRaw) {
                $searchAppSite = trim($searchAppSiteRaw);
                if (empty($searchAppSite)) continue;
                $searchTemplates = self::loadTemplatesFromSingleAppSite($rootDirPath, $searchAppSite);
                foreach ($searchTemplates as $k => $v) {
                    if (!isset($result[$k])) {
                        $result[$k] = $v;
                        Logger::debug("Added fallback template '$k' from '$searchAppSite'", 'LoaderNormal');
                    }
                }
            }
        }

        self::$htmlTemplatesCache[$cacheKey] = $result;
        return $result;
    }

    // Helper to load templates from a single AppSite (no caching/fallback)
    private static function loadTemplatesFromSingleAppSite(string $rootDirPath, string $appSite): array
    {
        $result = [];
        $appSitesPath = $rootDirPath . DIRECTORY_SEPARATOR . 'AppSites' . DIRECTORY_SEPARATOR . $appSite;
        if (!is_dir($appSitesPath)) {
            Logger::warn("AppSites directory not found: $appSitesPath", 'LoaderNormal');
            return $result;
        }
        Logger::debug("Loading templates from: $appSitesPath", 'LoaderNormal');
        $iterator = new \RecursiveIteratorIterator(
            new \RecursiveDirectoryIterator($appSitesPath, \RecursiveDirectoryIterator::SKIP_DOTS),
            \RecursiveIteratorIterator::LEAVES_ONLY
        );
        foreach ($iterator as $file) {
            if ($file->isFile() && $file->getExtension() === 'html') {
                $fileName = $file->getBasename('.html');
                $key = strtolower($appSite) . '_' . strtolower($fileName);
                $htmlContent = CommonUtil::normalizeFileContent(file_get_contents($file->getPathname()));
                Logger::debug("Loading template: $key (html size: " . strlen($htmlContent ?: '') . ")", 'LoaderNormal');
                // Find JSON file case-insensitively
                $jsonFile = substr($file->getPathname(), 0, -5) . '.json';
                $jsonContent = null;
                if (file_exists($jsonFile)) {
                    $jsonContent = CommonUtil::normalizeFileContent(file_get_contents($jsonFile));
                    Logger::debug("Found JSON file for $key (size: " . strlen($jsonContent ?: '') . ")", 'LoaderNormal');
                } else {
                    // Try case-insensitive search in the same directory
                    $dir = dirname($file->getPathname());
                    $baseName = strtolower($file->getBasename('.html'));
                    $files = scandir($dir);
                    if ($files !== false) {
                        foreach ($files as $entry) {
                            if ($entry !== '.' && $entry !== '..' && is_file($dir . DIRECTORY_SEPARATOR . $entry)) {
                                if (strtolower(pathinfo($entry, PATHINFO_EXTENSION)) === 'json') {
                                    $entryBase = strtolower(pathinfo($entry, PATHINFO_FILENAME));
                                    if ($entryBase === $baseName) {
                                        $matchedJsonPath = $dir . DIRECTORY_SEPARATOR . $entry;
                                        $jsonContent = CommonUtil::normalizeFileContent(file_get_contents($matchedJsonPath));
                                        Logger::debug("Found JSON file (case-insensitive) for $key (size: " . strlen($jsonContent ?: '') . ")", 'LoaderNormal');
                                        break;
                                    }
                                }
                            }
                        }
                    }
                }
                $result[$key] = new TemplateResult($htmlContent ?: '', $jsonContent ?: null);
            }
        }
        Logger::debug("Loaded " . count($result) . " templates for $appSite", 'LoaderNormal');
        return $result;
    }

    /**
     * Clear all cached templates (useful for testing or when templates change)
     */
    public static function clearCache(): void
    {
        self::$htmlTemplatesCache = [];
    }
}
?>

File: csharp/Assembler/Loader/LoaderNormal.cs
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.IO;
using Arshu.Common;
using Assembler.Common;

namespace Assembler.Loader;
// Always use Base JsonObject/JsonArray types for consistent processing
/// <summary>
/// Handles loading and caching of HTML templates from the file system
/// </summary>
public static class LoaderNormal
{
    #region Loading Templates

    private static readonly ConcurrentDictionary<string, Dictionary<string, (string html, string? json)>> _htmlTemplatesCache = new();

    /// <summary>
    /// Loads HTML files and corresponding JSON files from the specified application site directory, caching the output per appSite
    /// </summary>
    /// <param name="rootDirPath">Root directory path containing AppSites folder</param>
    /// <param name="appSite">Primary AppSite name to load</param>
    /// <param name="searchAppSites">Comma-delimited string of AppSite names to search for fallback templates (can be empty string)</param>
    public static Dictionary<string, (string html, string? json)> LoadGetTemplateFiles(string rootDirPath, string appSite, string searchAppSites)
    {
        Logger.Debug($"LoadGetTemplateFiles called for appSite: {appSite}, searchAppSites: {searchAppSites}", "LoaderNormal");

        var cacheKey = Path.GetDirectoryName(rootDirPath) + "|" + appSite + "|" + searchAppSites;
        if (_htmlTemplatesCache.TryGetValue(cacheKey, out var cached))
        {
            Logger.Debug($"Returning cached templates for {appSite} ({cached.Count} templates)", "LoaderNormal");
            return cached;
        }

        // Load templates from primary appSite
        var result = LoadTemplatesFromSingleAppSite(rootDirPath, appSite);

        // Load templates from searchAppSites for fallback
        if (!string.IsNullOrEmpty(searchAppSites))
        {
            var searchAppSitesArray = searchAppSites.Split(',');
            for (int i = 0; i < searchAppSitesArray.Length; i++)
            {
                var searchAppSite = searchAppSitesArray[i].Trim();
                if (string.IsNullOrEmpty(searchAppSite))
                    continue;

                var searchTemplates = LoadTemplatesFromSingleAppSite(rootDirPath, searchAppSite);
                foreach (var kvp in searchTemplates)
                {
                    // Only add if not already present (primary appSite takes precedence)
                    if (!result.ContainsKey(kvp.Key))
                    {
                        result[kvp.Key] = kvp.Value;
                        Logger.Debug($"Added fallback template '{kvp.Key}' from '{searchAppSite}'", "LoaderNormal");
                    }
                }
            }
        }

        _htmlTemplatesCache.TryAdd(cacheKey, result);
        return result;
    }

    /// <summary>
    /// Loads templates from a single AppSite without caching or fallback logic
    /// </summary>
    private static Dictionary<string, (string html, string? json)> LoadTemplatesFromSingleAppSite(string rootDirPath, string appSite)
    {
        var result = new Dictionary<string, (string html, string? json)>();
        var appSitesPath = Path.Combine(rootDirPath, "AppSites", appSite);

        if (!Directory.Exists(appSitesPath))
        {
            Logger.Warn($"AppSites directory not found: {appSitesPath}", "LoaderNormal");
            return result;
        }

        Logger.Debug($"Loading templates from: {appSitesPath}", "LoaderNormal");

        foreach (var file in Directory.GetFiles(appSitesPath, "*.html", SearchOption.AllDirectories))
        {
            var fileName = Path.GetFileNameWithoutExtension(file);
            var key = ($"{appSite.ToLowerInvariant()}_{fileName.ToLowerInvariant()}");
            var htmlContent = CommonUtil.NormalizeFileContent(File.ReadAllText(file));

            Logger.Debug($"Loading template: {key} (html size: {htmlContent.Length})", "LoaderNormal");

            // Find JSON file case-insensitively
            var jsonFile = Path.ChangeExtension(file, ".json");
            string? jsonContent = null;

            // Try exact match first
            if (File.Exists(jsonFile))
            {
                jsonContent = CommonUtil.NormalizeFileContent(File.ReadAllText(jsonFile));
                Logger.Debug($"Found JSON file for {key} (size: {jsonContent.Length})", "LoaderNormal");
            }
            else
            {
                // Try case-insensitive search in the same directory
                var directory = Path.GetDirectoryName(file);
                var baseFileName = Path.GetFileNameWithoutExtension(file);
                if (!string.IsNullOrEmpty(directory))
                {
                    var jsonFiles = Directory.GetFiles(directory, "*.json", SearchOption.TopDirectoryOnly);
                    string? matchingJson = null;
                    for (int i = 0; i < jsonFiles.Length; i++)
                    {
                        if (string.Equals(Path.GetFileNameWithoutExtension(jsonFiles[i]), baseFileName, StringComparison.OrdinalIgnoreCase))
                        {
                            matchingJson = jsonFiles[i];
                            break;
                        }
                    }

                    if (matchingJson != null)
                    {
                        jsonContent = CommonUtil.NormalizeFileContent(File.ReadAllText(matchingJson));
                        Logger.Debug($"Found JSON file (case-insensitive) for {key} (size: {jsonContent.Length})", "LoaderNormal");
                    }
                }
            }

            result[key] = (htmlContent, jsonContent);
        }

        Logger.Debug($"Loaded {result.Count} templates for {appSite}", "LoaderNormal");
        return result;
    }

    /// <summary>
    /// Clear all cached templates (useful for testing or when templates change)
    /// </summary>
    public static void ClearCache()
    {
        _htmlTemplatesCache.Clear();
    }

    #endregion
}